subscriber_notifee.go

```diff
@@ -14,82 +16,197 @@ import (
+			handlePeerIdentificationCompletedEvent(dht, evt)
+		case evt, more := <-nn.updateSub.Out():
+			if !more {
 				return
 			}
-
-			dht.plk.Lock()
-			if dht.host.Network().Connectedness(ev.Peer) != network.Connected {
-				dht.plk.Unlock()
-				continue
-			}
-
-			// if the peer supports the DHT protocol, add it to our RT and kick a refresh if needed
-			protos, err := dht.peerstore.SupportsProtocols(ev.Peer, dht.protocolStrs()...)
-			if err == nil && len(protos) != 0 {
-				refresh := dht.routingTable.Size() <= minRTRefreshThreshold
-				dht.Update(dht.ctx, ev.Peer)
-				if refresh && dht.autoRefresh {
-					select {
-					case dht.triggerRtRefresh <- nil:
-					default:
-					}
-				}
+			handlePeerProtocolsUpdatedEvent(dht, evt)
+		case evt, more := <-routabilityEvts:
+			if !more {
+				return
 			}
-			dht.plk.Unlock()
-
+			handleRoutabilityChanges(dht, evt)
 		case <-proc.Closing():
 			return
 		}
 	}
 }
 
+func handlePeerIdentificationCompletedEvent(dht *IpfsDHT, evt interface{}) {
+	// something has gone really wrong if we get an event for another type
+	e, ok := evt.(event.EvtPeerIdentificationCompleted)
+	if !ok {
+		logger.Errorf("got wrong type from subscription: %T", e)
+		return
+	}
+
+	dht.plk.Lock()
+	defer dht.plk.Unlock()
+	if dht.host.Network().Connectedness(e.Peer) != network.Connected {
+		return
+	}
+
+	// if the peer supports the DHT protocol, add it to our RT and kick a refresh if needed
+	protos, err := dht.peerstore.SupportsProtocols(e.Peer, dht.protocolStrs()...)
+	if err != nil {
+		logger.Errorf("could not check peerstore for protocol support: err: %s", err)
+		return
+	}
+	if len(protos) != 0 {
+		dht.Update(dht.ctx, e.Peer)
+		fixLowPeers(dht)
+	}
+}
+
+func handlePeerProtocolsUpdatedEvent(dht *IpfsDHT, evt interface{}) {
+	// something has gone really wrong if we get an event for another type
+	e, ok := evt.(event.EvtPeerProtocolsUpdated)
+	if !ok {
+		logger.Errorf("got wrong type from subscription: %T", evt)
+		return
+	}
+
+	protos, err := dht.peerstore.SupportsProtocols(e.Peer, dht.protocolStrs()...)
+	if err != nil {
+		logger.Errorf("could not check peerstore for protocol support: err: %s", err)
+		return
+	}
+
+	if len(protos) > 0 {
+		dht.routingTable.Update(e.Peer)
+	} else {
+		dht.routingTable.Remove(e.Peer)
+	}
+
+	fixLowPeers(dht)
+}
+
+func handleRoutabilityChanges(dht *IpfsDHT, evt interface{}) {
+	var target mode
+
+	switch evt.(type) {
+	case event.EvtLocalRoutabilityPrivate, event.EvtLocalRoutabilityUnknown:
+		target = modeClient
+	case event.EvtLocalRoutabilityPublic:
+		target = modeServer
+	}
+
+	logger.Infof("processed event %T; performing dht mode switch", evt)
+
+	err := dht.setMode(target)
+	// NOTE: the mode will be printed out as a decimal.
+	if err == nil {
+		logger.Infof("switched DHT mode successfully; new mode: %d", target)
+	} else {
+		logger.Warningf("switching DHT mode failed; new mode: %d, err: %s", target, err)
+	}
+}
+
+func fixLowPeers(dht *IpfsDHT) {
```

Comment:
Noticed that some of these functions had a check or another regarding really small routing tables, figured we might want to consolidate. Thoughts @aarshkshah1992 @Stebalien ?